### ISSUE TITLE
[Bug]: A question about the isTSX option

### ISSUE THREAD

**Issue by likaiqiang** [babel/babel#16285]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

babel-loader (webpack)

### Input code

```javascript
rules: [
            {
                test: /\.(?:js|mjs|cjs|jsx|ts|tsx)$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        plugins: [
                            [
                                '@babel/plugin-transform-typescript'
                            ],
                            [
                                '@babel/plugin-transform-react-jsx',
                                {
                                    runtime: 'automatic',
                                    importSource: 'fre',
                                },
                            ]
                        ]
                    }
                }
            }
        ],
```

### Current and expected behavior

At https://github.com/babel/babel/blob/v7.23.10/packages/babel-plugin-syntax-typescript/src/index.ts#L59, the isTSX option is used to determine whether to enable the jsx plugin, but at https://github.com/babel/babel/blob/v7.23.10/packages/babel-plugin-syntax-jsx/src/index.ts#L19, it only checks whether there is typescript in the plugins, and if there is, it returns. Is this not quite right, babel-plugin-syntax-jsx should also check whether there is isTSX, so should the 19th line be written like this?

```javascript
if (parserOpts.plugins.some(p => (Array.isArray(p) ? p[0] : p) === "typescript") && parserOpts.plugins.includes('jsx')) {
  return;
}
```

### Environment

System:
    OS: Windows 11 10.0.22621
  Binaries:
    Node: 18.16.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 9.5.1 - C:\Program Files\nodejs\npm.CMD
    pnpm: 7.30.0 - ~\AppData\Local\pnpm\pnpm.EXE
  npmPackages:
    @babel/plugin-transform-react-jsx: ^7.23.4 => 7.23.4 
    @babel/plugin-transform-typescript: ^7.23.5 => 7.23.5
    @babel/preset-env: ^7.23.5 => 7.23.5
    babel-loader: ^9.1.3 => 9.1.3
    webpack: ^5.89.0 => 5.89.0

**Comment by nicolo-ribaudo** [comment]
Hey! Sorry it took so long to answer.

We found that it was easier to completely control TSX within the TS plugin, since for many TS users TS supports JSX out of the box and they don't really thing about it as a separate language extension.

When using JSX with TypeScript, by default the TS plugin will:
- enable TS and JSX on `.tsx` files
- enable TS and disable JSX on `.ts` files

We decided as a "policy" that plugins are only configured through explicit options and not implicitly through the filename, so `isTSX` means "this file has a .tsx extension".

The JSX plugin checks only for `typescript` because these are three possible cases:
- `typescript` is present. This is either a `.ts` or ` .tsx` file, and the TS plugin will already have enabled/disabled JSX support as needed
- `typescript` is not present:
  - the typescript plugin is not involved at all. Just enable JSX parsing.
  - the typescript plugin will be applied after this one. It will take care of removing the JSX parser option if it's not a TSX file.
